streams/shared/src/main/scala/zio/stream/ZStreamChunk.scala

```diff
@@ -89,6 +89,52 @@ class ZStreamChunk[-R, +E, +A](val chunks: ZStream[R, E, Chunk[A]]) extends Seri
   final def catchAllCause[R1 <: R, E2, A1 >: A](f: Cause[E] => ZStreamChunk[R1, E2, A1]): ZStreamChunk[R1, E2, A1] =
     ZStreamChunk(chunks.catchAllCause(c => f(c).chunks))
 
+  /**
+   * Chunks the stream with specifed chunkSize
+   * @param chunkSize size of the chunk
+   */
+  final def chunkN(chunkSize: Int): ZStreamChunk[R, E, A] =
+    ZStreamChunk {
+      ZStream {
+        self.process.mapM { as =>
+          Ref.make[Option[Option[E]]](None).flatMap { stateRef =>
```

Comment:
Given that the result must fail on the first error, but only produce a chunk whenever the current pull reaches `n` or the end of the stream, I wonder if something like this wouldn't be clearer:
```scala
ZStreamChunk(ZStream {
  chunks.process.mapM { xss =>
    Ref.make[(Boolean, Chunk[A])](false -> Chunk.empty).map { state =>
      def pull: Pull[R, E, Chunk[A]] =
        state.get.flatMap {
          case (true, _) => Pull.end
      
          case (false, chunk) if chunk.length >= chunkSize =>
                val m = Math.min(chunk.length, chunkSize)
                Pull.emit(chunk.take(m)) <* state.set(false -> chunk.drop(m))

          case (false, chunk0) =>  
                xss.foldM({
                  case None        => Pull.emit(chunk0) <* state.set(true -> Chunk.empty)
                  case e @ Some(_) => ZIO.fail(e)
                }, 
                xs => state.modify { case (_, chunk) =>
                  if (chunk.length + xs.length >= chunkSize) {
                    val m = chunkSize - chunk.length
                    Pull.emit(chunk ++ xs.take(m)) -> (false -> xs.drop(m))
                  } else
                    pull -> (false -> (chunk ++ xs))
                }.flatten)
            }
            
      pull
    }
  }
})
```

Reply:
Thank you for your suggestion, this design makes better sense to me. If you don't mind, I'd like to refactor the code using this as base.

Reply:
Sure thing !